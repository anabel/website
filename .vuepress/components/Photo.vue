<template>
  <picture v-if="breakpoints">
    <source
      v-for="breakpoint in this.breakpoints"
      :srcset="getImageUrl(src(breakpoint))"
      :media="media(breakpoint)"
    />
    <img :src="getImageUrl(this.name)" :alt="this.alt" />
  </picture>
  <img v-else :src="getImageUrl(this.name)" :alt="this.alt" />
</template>

<script>
const mediaCondition = "max-width";
const separator = ".";

export default {
  name: "photo",
  props: {
    name: {
      type: String,
      required: true
    },
    alt: {
      type: String
    },
    breakpoints: {
      type: Array
    }
  },
  methods: {
    media(breakpoint) {
      return "("
        .concat(mediaCondition)
        .concat(": ")
        .concat(this.$mediaQueries.get(breakpoint))
        .concat(")");
    },
    src(breakpoint) {
      if (!this.name.includes(separator)) {
        return this.name;
      }
      let filename = this.name.split(separator);
      return [filename[0], breakpoint]
        .join("_")
        .concat(separator)
        .concat(filename[1]);
    },
    getImageUrl(filename) {
      try {
        return require("@images/" + filename);
      } catch (error) {}
    }
  }
};
</script>