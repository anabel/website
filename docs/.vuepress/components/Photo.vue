<template>
  <picture v-if="breakpoints">
    <source
      v-for="breakpoint in this.breakpoints"
      :srcset="getImageUrl(src(breakpoint))"
      :media="media(breakpoint)"
    />
    <img :src="getImageUrl(this.name)" :alt="this.alt" />
  </picture>
  <img v-else :src="getImageUrl(this.name)" :alt="this.alt" />
</template>

<script>
import { mediaQueries } from "@dynamic/tailwindMediaQueries.js";

export default {
  data() {
    return {
      mediaQueries: mediaQueries
    };
  },
  props: {
    name: {
      type: String,
      required: true
    },
    alt: {
      type: String
    },
    breakpoints: {
      type: Array
    }
  },
  computed: {
    mediaConditions() {
      let conditions = new Map();
      for (const screen in mediaQueries) {
        if (mediaQueries.hasOwnProperty(screen)) {
          let condition = this.mediaCondition(mediaQueries[screen]);
          conditions.set(screen, condition);
        }
      }
      return conditions;
    }
  },
  methods: {
    media(breakpoint) {
      return this.mediaConditions.get(breakpoint);
    },
    src(breakpoint) {
      const separator = ".";
      if (!this.name.includes(separator)) {
        return this.name;
      }
      let filename = this.name.split(separator);
      return [filename[0], breakpoint]
        .join("_")
        .concat(separator)
        .concat(filename[1]);
    },
    getImageUrl(filename) {
      try {
        return require("@images/" + filename);
      } catch (error) {}
    },
    mediaCondition(mediaQuery) {
      const mediaFeature = "width";
      let mediaCondition = [];
      for (const condition in mediaQuery) {
        let media;
        if (mediaQuery.hasOwnProperty(condition)) {
          let value = mediaQuery[condition];
          media = "("
            .concat(condition)
            .concat("-")
            .concat(mediaFeature)
            .concat(": ")
            .concat(value)
            .concat(")");
        }
        mediaCondition.push(media);
      }
      return mediaCondition.join(" and ");
    }
  }
};
</script>