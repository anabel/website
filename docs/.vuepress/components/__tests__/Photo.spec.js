import Photo from "../Photo.vue";
import { RouterLinkStub, shallowMount } from "@vue/test-utils";

jest.mock("../twMediaQueryAdapter", () => ({
  get mediaConditions() {
    let mediaConditions = new Map();
    mediaConditions.set("sm", "(min-width: 640px) and (max-width: 767px)");
    return mediaConditions;
  }
}));

jest.mock("../urlUtils.js", () => ({
  url: jest.fn(() => "/path/foto.jpg")
}));

describe("If no breakpoints are passed", () => {
  const wrapper = shallowMount(Photo, {
    propsData: {
      name: "foto.jpg",
      alt: "descripcion de la foto"
    }
  });

  it("source tags are not displayed", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe("If a breakpoint is passed", () => {
  const wrapper = shallowMount(Photo, {
    propsData: {
      name: "foto.jpg",
      alt: "descripcion de la foto",
      breakpoints: ["sm"]
    }
  });

  it("source tags is displayed with corresponding media condition", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
