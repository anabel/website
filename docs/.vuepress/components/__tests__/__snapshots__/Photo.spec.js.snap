// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`If a breakpoint is passed source tags is displayed with corresponding filename and media condition 1`] = `
<picture>
  <source srcset="/path/foto_sm.jpg" media="(min-width: 640px) and (max-width: 767px)"> <img src="/path/foto.jpg" alt="descripcion de la foto">
</picture>
`;

exports[`If a non existent breakpoint is passed source tags is not displayed 1`] = `<picture> <img src="/path/foto.jpg" alt="descripcion de la foto"></picture>`;

exports[`If multiple breakpoints are passed only valid breakpoints are displayed 1`] = `
<picture>
  <source srcset="/path/foto_sm.jpg" media="(min-width: 640px) and (max-width: 767px)"> <img src="/path/foto.jpg" alt="descripcion de la foto">
</picture>
`;

exports[`If no breakpoints are passed source tags are not displayed 1`] = `<picture> <img src="/path/foto.jpg" alt="descripcion de la foto"></picture>`;
